File: src/Emerald/Ast.cs
namespace Emerald;

/// <summary>
/// An error stops the program; a warning does not. §2.6 makes warnings a design surface
/// rather than an afterthought — the compiler is the most patient teacher in the room —
/// which is only possible once a diagnostic can be something other than fatal.
/// </summary>
public enum Severity { Error, Warning }

public sealed record Diagnostic(
    string File, int Line, string Message, string? Hint = null,
    Severity Severity = Severity.Error,

    /// <summary>Which worked example `emerald explain` should show for this. Null where
    /// none is written yet, which is most of them and is said out loud rather than
    /// papered over.</summary>
    string? Topic = null);

/// <summary>
/// The written shape of a callable: <c>func(Int): String</c>. A function's type is its
/// header with the name removed, so there is no second grammar to learn — and the return
/// type is optional here for the same reason it is optional on a declaration.
/// </summary>
public sealed record FuncRef(List<TypeRef> Params, TypeRef? Returns);

/// <summary>
/// A written type. <c>Arguments</c> holds the type arguments of a compiler-owned generic;
/// <c>Function</c> is set instead when the type is a callable, and then <c>Name</c> is the
/// <c>func</c> keyword itself, kept only so a diagnostic has a line to point at.
/// </summary>
public sealed record TypeRef(Token Name, bool Nullable, List<TypeRef>? Arguments = null,
                             FuncRef? Function = null);

public sealed record Param(Token Name, TypeRef? Type, Expr? Default);

/// <summary>
/// <c>@export</c>, <c>@name("Any")</c> — declarative, compiler-known, and generating no
/// code (§3.8). Named Attr rather than Attribute because .NET's implicit usings already
/// bring System.Attribute into scope, and two things called Attribute in one file is a
/// puzzle nobody needs to solve twice.
/// </summary>
public sealed record Attr(Token Name, Expr? Argument);

/// <summary>One <c>key: value</c> pair of a dictionary literal.</summary>
public sealed record Entry(Expr Key, Expr Value);

// ---- expressions --------------------------------------------------------

public abstract record Expr
{
    /// <summary>
    /// A constant. <c>Line</c> exists only so a diagnostic can point at one: an expression
    /// made entirely of literals — <c>if true then "a" else 42</c> — otherwise carried no
    /// token anywhere, and the error came out at line 0 with no source line to quote.
    /// </summary>
    public sealed record Literal(object? Value, int Line = 0) : Expr;

    /// <summary>"a #{b} c" — alternating literal and expression parts.</summary>
    public sealed record Interpolation(List<Expr> Parts) : Expr;

    public sealed record Variable(Token Name) : Expr;
    public sealed record Grouping(Expr Inner) : Expr;
    public sealed record Unary(Token Op, Expr Right) : Expr;
    public sealed record Binary(Expr Left, Token Op, Expr Right) : Expr;

    /// <summary>
    /// <c>value is Dog</c>. Asks what a value actually is at runtime, and narrows the
    /// name to that type for the branch where the answer is yes — the same machinery
    /// the <c>nothing</c> check has always used, pointed at a second question.
    /// </summary>
    public sealed record TypeTest(Expr Value, Token Keyword, TypeRef Type) : Expr;

    /// <summary>
    /// <c>animal as Dog</c>, giving <c>Dog?</c>. The same question <c>is</c> asks, answered
    /// as a value rather than as a narrowing — for the places narrowing cannot reach, since
    /// it is recorded against a name and a field is not one.
    /// </summary>
    public sealed record TypeCast(Expr Value, Token Keyword, TypeRef Type) : Expr;

    /// <summary>Short-circuits, so it cannot share Binary's evaluation.</summary>
    public sealed record Logical(Expr Left, Token Op, Expr Right) : Expr;

    /// <summary>if c then a else b — the expression form. `else` is mandatory (§3.1).</summary>
    public sealed record IfExpr(Expr Condition, Expr Then, Expr Else) : Expr;

    public sealed record RangeExpr(Expr Start, Expr End) : Expr;

    /// <summary>[1, 2, 3]</summary>
    public sealed record ListLiteral(Token Bracket, List<Expr> Items) : Expr;

    /// <summary>
    /// <c>["a": 1, "b": 2]</c>, and <c>[:]</c> for an empty one. Shares the bracket with
    /// a list rather than taking <c>{ }</c>, which is a block and a trailing lambda here —
    /// <c>{ x: 1 }</c> could not be told from <c>{ x =&gt; 1 }</c> without lookahead nobody
    /// should have to do. Swift solves it the same way, for the same reason.
    /// </summary>
    public sealed record DictLiteral(Token Bracket, List<Entry> Entries) : Expr;

    /// <summary>a[0] — lowered through the Indexable trait once traits exist (§3.2).</summary>
    public sealed record Index(Expr Target, Token Bracket, Expr Position) : Expr;
    public sealed record Lambda(List<Param> Params, List<Stmt> Body) : Expr;

    /// <summary>
    /// Property or method access: <c>dog.name</c>, <c>5.times</c>. <c>Optional</c> is the
    /// <c>?.</c> form, where a receiver that is nothing gives nothing back unread.
    /// </summary>
    public sealed record Get(Expr Target, Token Name, bool Optional = false) : Expr;

    /// <summary>
    /// A call. <c>Trailing</c> is the trailing-lambda sprinkle: <c>1.upto(5) { i => ... }</c>
    /// and <c>5.times { ... }</c> both land here.
    /// </summary>
    /// <summary>
    /// <c>TypeArgs</c> is the <c>&lt;Dog&gt;</c> of <c>animal.as&lt;Dog&gt;()</c>: type
    /// arguments written at a call site. Consuming a generic method, never declaring one
    /// — §5.3 defers declaring, and nothing here introduces a type parameter.
    /// </summary>
    public sealed record Call(
        Expr Callee, List<Expr> Args, Lambda? Trailing,
        List<TypeRef>? TypeArgs = null) : Expr;
}

// ---- statements ---------------------------------------------------------

public abstract record Stmt
{
    public sealed record VarDecl(
        Token Name, TypeRef? Type, Expr? Init, bool IsConst,
        bool IsStatic = false,
        List<Stmt>? Getter = null, List<Stmt>? Setter = null,
        List<Attr>? Attributes = null) : Stmt;

    /// <summary>Assignment is a statement, never an expression (§3.1) — so
    /// <c>if x = 5 { }</c> cannot parse, and the =/== bug is unrepresentable.</summary>
    public sealed record Assign(Expr Target, Token Op, Expr Value) : Stmt;

    public sealed record ExprStmt(Expr Expression) : Stmt;

    /// <summary>`else if` is represented as an Else branch holding a single If.</summary>
    public sealed record If(Expr Condition, List<Stmt> Then, List<Stmt>? Else) : Stmt;

    public sealed record While(Expr Condition, List<Stmt> Body) : Stmt;
    /// <summary>
    /// <c>for x in xs</c>, and <c>for (k, v) in pairs</c> when <c>Second</c> is present.
    /// A pair is worth having only if it can be taken apart, and the loop is where taking
    /// one apart is most often wanted.
    /// </summary>
    public sealed record For(
        Token Variable, Expr Iterable, List<Stmt> Body, Token? Second = null) : Stmt;

    /// <summary>
    /// <c>var (name, score) = best</c>. Two names bound from one pair.
    ///
    /// Its own statement rather than a VarDecl carrying a list, because almost everything
    /// a VarDecl can be -- a property with a get body, a static, an annotated field --
    /// is meaningless here, and folding it in would mean writing "not on a destructure"
    /// into a dozen checks that currently do not have to think about it.
    /// </summary>
    public sealed record PairDecl(
        Token First, Token Second, Expr Init, bool IsConst) : Stmt;
    /// <summary>A null <c>Body</c> means abstract: required, not provided (§3.2).</summary>
    public sealed record FuncDecl(
        Token Name, List<Param> Params, TypeRef? ReturnType, List<Stmt>? Body,
        bool IsStatic = false, List<Attr>? Attributes = null,

        /// <summary>Declared with the <c>override</c> keyword — a claim that this replaces
        /// an inherited implementation, checked in both directions (§3.2).</summary>
        bool IsOverride = false) : Stmt;
    public sealed record Return(Token Keyword, Expr? Value) : Stmt;

    /// <summary>
    /// Leaving a loop early, and skipping to its next turn. Both carry their keyword
    /// token only — there is no labeled form, so there is nothing else to record.
    /// </summary>
    public sealed record Break(Token Keyword) : Stmt;
    public sealed record Continue(Token Keyword) : Stmt;

    public sealed record Throw(Token Keyword, Expr Value) : Stmt;

    /// <summary>
    /// <c>assert total == 10</c>. A statement rather than a call, so it needs no
    /// parentheses — the same shape as <c>throw</c>, and what §3.5 already wrote.
    ///
    /// It is the one construct that reads its own argument as <em>source</em> rather than
    /// as a value: a function receives <c>false</c> and can say nothing else, which is why
    /// §3.8 filed this as the only real macro demand.
    /// </summary>
    public sealed record Assert(Token Keyword, Expr Condition) : Stmt;

    /// <summary>
    /// <c>try { } catch e { }</c>. The catch name is bound to an Error inside the handler.
    /// Untyped in v0 — catching by type needs user-declared error classes first.
    /// </summary>
    public sealed record TryCatch(
        Token Keyword, List<Stmt> Body, Token CaughtName, List<Stmt> Handler) : Stmt;

    /// <summary>
    /// A class. <c>Members</c> holds VarDecls (fields), FuncDecls (methods), and at most
    /// one Constructor. The block-free file form (§3.3) produces the same node — only the
    /// parsing differs, so nothing downstream needs to know which shape was written.
    /// </summary>
    /// <summary>
    /// <c>Initializer</c> holds a module file's top-level statements — the code that is not
    /// a declaration. §3.3 lowers them to a static constructor, run once on first member
    /// access. They used to be handed through as "members", where every pass that walked
    /// members ignored anything that was not a declaration, and they ran nowhere at all.
    /// </summary>
    public sealed record ClassDecl(
        TypeKind Kind, Token Name, Token? BaseName,
        List<Token> Traits, List<Stmt> Members,
        List<Attr>? Attributes = null, List<Stmt>? Initializer = null,

        /// <summary>A file that declared no type, wrapped into one by <see cref="Project"/>.
        /// It reads like a class everywhere else, but its members see each other by their
        /// bare names — a file's own contents were the one thing it could not (§3.3).</summary>
        bool IsModule = false) : Stmt;

    public sealed record ConstructorDecl(
        Token Keyword, List<Param> Params, List<Stmt> Body) : Stmt;

    /// <summary>
    /// <c>enum Color { RED, GREEN, BLUE }</c> — a closed set of named values, and
    /// nothing else. No payloads: §6 harvested the demand as alignment, dock, orientation
    /// and color, all of which are plain names, and a tagged union is a different feature
    /// wearing the same keyword.
    /// </summary>
    public sealed record EnumDecl(
        Token Name, List<Token> Members, List<Attr>? Attributes = null,

        /// <summary>Methods written after the values. Payloads are still out — a tagged
        /// union is a different feature — but a fact about a value belongs on it (§3.2).</summary>
        List<Stmt>? Methods = null) : Stmt;
}

/// <summary>
/// Class, trait, and struct share one node — they differ in what they may contain and
/// how they are used, not in how they are written or parsed.
/// </summary>
public enum TypeKind { Class, Trait, Struct, Enum }

File: src/Ast.zig
//! The syntax tree.
//!
//! Section 19.2 keeps this free of runtime values, so the tree a future backend
//! consumes is the same tree the interpreter consumes. Every node carries the
//! span it came from, which diagnostics, stack traces, and later the formatter
//! and language server all read.

const std = @import("std");
const Source = @import("Source.zig");

/// A whole source file.
pub const Program = struct {
    statements: []const Statement,
    /// Section 14.2's `using` declarations. They are file-local and name no
    /// order of execution, so they are kept beside the statements rather than
    /// among them.
    using: []const Using = &.{},
};

/// Section 14.2: `using Shapes`, which makes that namespace's public names
/// directly visible here, or `using UiColor = Graphics.Color`, which gives one
/// name a short spelling. Either way it is file-local, imports only direct
/// public names, and neither includes nor executes anything.
pub const Using = struct {
    span: Source.Span,
    /// The short name this introduces, empty for the unaliased form, where
    /// every public name of `path` keeps its own spelling.
    alias: []const u8 = "",
    alias_span: Source.Span = .{ .start = 0, .end = 0 },
    /// The dotted path as written: `Graphics.Color` is `.{ "Graphics", "Color" }`.
    path: []const []const u8,
    path_span: Source.Span,
};

/// A brace-delimited sequence of statements. Section 6.1 gives every one its own
/// lexical scope.
pub const Block = struct {
    span: Source.Span,
    statements: []const Statement,
};

pub const Statement = struct {
    span: Source.Span,
    data: Data,

    pub const Data = union(enum) {
        /// An expression evaluated for its effect. Section 5.2 allows this only
        /// for calls; a pure expression whose result is unused is an error.
        expression: *const Expression,
        declaration: Declaration,
        assignment: Assignment,
        conditional: If,
        while_loop: While,
        for_loop: For,
        /// Section 6.4's `break`, holding its keyword's span.
        break_statement: Source.Span,
        /// Section 6.4's `continue`, holding its keyword's span.
        continue_statement: Source.Span,
        function_declaration: FunctionDeclaration,
        /// Section 10's user-defined value type and its stored fields.
        struct_declaration: StructDeclaration,
        return_statement: Return,
        /// Section 8.2's `var (name, age) = entry`.
        destructuring: Destructuring,
        /// Section 8.2's `(left, right) = (right, left)`, which assigns to
        /// names that already exist.
        destructuring_assignment: DestructuringAssignment,
    };
};

pub const StructDeclaration = struct {
    /// Section 10.1: a class declares a reference type with the same members a
    /// struct has. Everything that differs follows from sharing.
    class: bool = false,
    /// Section 11.1: a trait declares a contract, whose members are
    /// requirements or defaults, and stores nothing.
    trait: bool = false,
    name: []const u8,
    name_span: Source.Span,
    /// Section 10.7's `extends Animal`: the one base class a class may have.
    base: ?TypeExpression = null,
    /// Section 11.2's `with Swimmer, Flyer`: the traits adopted, or for a
    /// trait, the ones it builds on.
    traits: []const TypeExpression = &.{},
    /// Section 10.7's `@abstract`, which keeps the class from being constructed
    /// and lets its methods leave out their bodies.
    abstract_span: ?Source.Span = null,
    fields: []const Field,
    /// Section 10.2's custom constructor, which replaces the generated one.
    /// A type declares at most one, since overloading is deferred.
    constructor: ?Constructor = null,
    /// Section 10's instance methods. Each sees the instance as `self`.
    methods: []const FunctionDeclaration = &.{},
    /// Section 10.3's computed properties.
    properties: []const Property = &.{},
    /// Section 10.4's `func Vector2.origin()`, which belongs to the type
    /// rather than to each value, and so has no `self`.
    type_functions: []const TypeFunction = &.{},
    /// Section 10.4's `var Player.count = 0`.
    type_fields: []const TypeField = &.{},

    /// The keyword it was declared with, for diagnostics.
    pub fn keyword(self: StructDeclaration) []const u8 {
        return if (self.trait) "trait" else if (self.class) "class" else "struct";
    }

    /// The declaration's own `name` is the whole `Vector2.origin` as written,
    /// which is how a stack trace or a diagnostic about the function reads;
    /// `member` is the part after the dot, which shares the type's member
    /// name space.
    pub const TypeFunction = struct {
        member: []const u8,
        member_span: Source.Span,
        declaration: FunctionDeclaration,
    };

    pub const TypeField = struct {
        mutable: bool,
        name: []const u8,
        name_span: Source.Span,
        /// Optional, as for a module-level binding: the type can come from
        /// the value.
        annotation: ?TypeExpression,
        /// Required: section 10.4's type-level fields "require initial
        /// values", since nothing else runs to assign one.
        initializer: *const Expression,
    };

    pub const Field = struct {
        mutable: bool,
        name: []const u8,
        name_span: Source.Span,
        annotation: TypeExpression,
        /// Section 10.2's default, which runs when construction leaves the
        /// field to it.
        default: ?*const Expression = null,
    };

    /// `const area: Float { ... }`, or `var diameter: Float { get { ... } set
    /// { ... } }`. The parser builds each accessor as an ordinary method
    /// declaration — a getter with no parameters returning the property's
    /// type, a setter taking `value` — so every later pass calls them the way
    /// it calls any method.
    pub const Property = struct {
        mutable: bool,
        name: []const u8,
        name_span: Source.Span,
        annotation: TypeExpression,
        getter: FunctionDeclaration,
        setter: ?FunctionDeclaration,
        /// Section 10.7's `@override`, which replaces a base class's property.
        override_span: ?Source.Span = null,
    };

    /// `constructor(x: Float) { self.x = x }`. It has no name and no return
    /// type: calling the type runs it, and what it produces is always `self`.
    pub const Constructor = struct {
        keyword_span: Source.Span,
        parameters: []const Parameter,
        body: Block,
    };
};

/// Section 8.2's `(name, age)`: the names a tuple is unpacked into. `_`
/// discards its position, exactly as it does elsewhere. A position may itself
/// be a pattern, as in `(id, (x, y))` (7.4).
pub const Pattern = struct {
    span: Source.Span,
    /// One per position of the tuple, in order.
    positions: []const Position,
    /// Every name the pattern binds, nested ones included, in the order
    /// written. What only needs the names, such as declaring them, reads this.
    names: []const Name,

    pub const Name = struct {
        text: []const u8,
        span: Source.Span,
    };

    pub const Position = union(enum) {
        name: Name,
        nested: *const Pattern,
    };
};

/// `var (name, age) = entry`, or `const` (4.3). An annotation applies to the
/// whole tuple, not to one position.
pub const Destructuring = struct {
    mutable: bool,
    pattern: Pattern,
    annotation: ?TypeExpression,
    initializer: *const Expression,
};

/// `(left, right) = (right, left)`. Section 8.2 evaluates the complete right
/// side before any destination changes, which is what makes a swap work.
pub const DestructuringAssignment = struct {
    pattern: Pattern,
    value: *const Expression,
};

/// Section 6.4: `while condition { body }`.
pub const While = struct {
    condition: *const Expression,
    body: Block,
};

/// Section 6.4: `for name in iterable { body }`. The binding is read-only and
/// fresh for every iteration. Only a range can be looped over so far;
/// collections arrive with the collection slice.
pub const For = struct {
    /// `_` discards each value without introducing a binding. Empty when the
    /// loop destructures instead, which section 8.2 allows wherever a binding
    /// is introduced.
    name: []const u8,
    name_span: Source.Span,
    /// Section 8.2's `for (name, age) in entries`; null for a plain binding.
    pattern: ?Pattern = null,
    iterable: *const Expression,
    body: Block,
};

/// Section 7.1. Parameters are read-only, and require an explicit type this
/// slice rather than the defaults section 7.2 allows for public API guidance.
/// Nested function declarations, defaults, and named arguments are deferred; a
/// name declares at most one function, per section 7.3. Anonymous functions are
/// `Expression.Lambda`.
pub const FunctionDeclaration = struct {
    name: []const u8,
    name_span: Source.Span,
    parameters: []const Parameter,
    /// Omitted for a function with no result, whose return type is then
    /// `Nothing`, exactly as if `: Nothing` were written (section 7.2).
    return_annotation: ?TypeExpression,
    body: Block,
    /// Section 10.7's `@override`, which replaces a base class's method.
    override_span: ?Source.Span = null,
    /// Section 10.7's `@abstract`, on a method with no body for a subclass to
    /// supply, or for a trait's method written without a body, which is a
    /// requirement (11.1), its name. Its `body` is empty.
    abstract_span: ?Source.Span = null,
};

pub const Parameter = struct {
    name: []const u8,
    name_span: Source.Span,
    annotation: TypeExpression,
    /// Section 7.3's default, evaluated when a call omits this parameter.
    default: ?*const Expression = null,
};

pub const Return = struct {
    keyword_span: Source.Span,
    /// Null for a bare `return`, which section 7.1 allows for a function with
    /// no result.
    value: ?*const Expression,
};

pub const Declaration = struct {
    /// Section 4.3: `var` permits rebinding, `const` does not.
    mutable: bool,
    name: []const u8,
    name_span: Source.Span,
    /// Section 4.1 infers the type from the initializer when there is no
    /// annotation, and requires an annotation when there is no initializer.
    annotation: ?TypeExpression,
    initializer: ?*const Expression,
};

/// A type as written in the source: a name, section 8.2's `[T]`, or section
/// 7.1's `func(Int): String`. The dictionary and set spellings arrive with
/// those collections.
pub const TypeExpression = struct {
    span: Source.Span,
    /// The name, or empty for a list or function type.
    name: []const u8,
    /// The element type of a list type, `T` in `[T]`, the value type of a
    /// dictionary, or the member type of a set; null otherwise.
    element: ?*const TypeExpression = null,
    /// The key type of a dictionary type, `K` in `[K: V]`; null otherwise.
    key: ?*const TypeExpression = null,
    /// Whether `element` is section 8.2's `{T}` rather than `[T]`.
    set: bool = false,
    /// The shape of a function type; null otherwise.
    signature: ?*const SignatureExpression = null,
    /// The position types of a tuple type, `(A, B)`; null otherwise.
    positions: ?[]const TypeExpression = null,
    /// The `?` that marks an optional, split from the name by the parser as
    /// section 4.2 describes. Null when the type is not optional.
    question_span: ?Source.Span,
};

/// `name = value`, or an assignment reached through a path of indices and
/// fields, `name[i].field[j] = value`.
pub const Assignment = struct {
    /// The binding assigned, or the one whose place is reached through
    /// `steps`.
    name: []const u8,
    name_span: Source.Span,
    /// The path from outermost to innermost, empty for a plain assignment.
    /// `grid[0][1] = 5` has two index steps; `point.x = 1` has one field step.
    steps: []const Step = &.{},
    /// The whole destination as written, for diagnostics.
    target_span: Source.Span,
    /// The operation a compound assignment applies, or null for a plain `=`.
    /// Section 5.3 lowers `a += b` through the same operation as `a + b`.
    operation: ?BinaryOperator,
    value: *const Expression,
};

/// One step of an assignment path: an index into a list or dictionary, or a
/// named struct field. `entry.0` never appears here — a tuple position can
/// never be assigned to, which the parser rejects before building one.
pub const Step = union(enum) {
    index: *const Expression,
    field: struct { name: []const u8, span: Source.Span },
};

/// `func(Int, String): Bool` as written. The result is null when the type
/// omits it, which section 7.1 makes the same as writing `Nothing`.
pub const SignatureExpression = struct {
    parameters: []const TypeExpression,
    result: ?*const TypeExpression,
};

pub const If = struct {
    condition: *const Expression,
    then_block: Block,
    otherwise: ?Else,
    /// Section 6.2's trailing form, `statement if condition`, which guards one
    /// statement and has no `else`. Its `then_block` holds that one statement.
    /// Kept distinct so the source shape survives for the formatter; nothing
    /// else treats it differently from a block `if`.
    trailing: bool = false,
};

pub const Else = union(enum) {
    block: Block,
    /// An `else if`, kept distinct from a block holding one `if` so the source
    /// shape survives for the formatter.
    chained: *const Statement,
};

pub const Expression = struct {
    span: Source.Span,
    data: Data,
    /// The height of this expression's tree. Every pass walks expressions
    /// recursively, so the parser bounds this to keep them all within the host
    /// stack; see `Parser.max_expression_depth`.
    depth: u32 = 1,

    pub const Data = union(enum) {
        int_literal: i64,
        float_literal: f64,
        bool_literal: bool,
        /// Section 4.2's `nothing`, the single value of the absence-only type.
        nothing_literal: void,
        name: []const u8,
        unary: Unary,
        binary: Binary,
        logical: Logical,
        comparison: Comparison,
        call: Call,
        range: Range,
        /// A string whose text is fully known, with every escape and, for a
        /// triple-quoted string, its indentation already applied (5.1).
        string_literal: []const u8,
        /// Section 5.1's `"text #{expression} text"`, in order.
        interpolation: []const Part,
        /// Section 8.2's `[a, b, c]`. A set literal shares this spelling, and
        /// is told apart by the type expected where it is written.
        list_literal: []const *const Expression,
        /// Section 8.2's `["Ava": 12]`, recognized by its `key: value` entries.
        dictionary_literal: []const Entry,
        /// Section 5.4's zero-based `base[index]`.
        index: Index,
        /// `base.name`: a property, or a method when it is the callee of a call.
        member: Member,
        /// Section 7.4's `{ value => value * 2 }`.
        lambda: Lambda,
        /// Section 8.2's `("score", 10)`, which always has at least two
        /// positions; one parenthesized expression is a group.
        tuple_literal: []const *const Expression,
        /// Section 4.4's `animal is Dog`.
        type_test: TypeTest,
    };

    /// `value is Type`, which asks what the value is at runtime.
    pub const TypeTest = struct {
        value: *const Expression,
        target: TypeExpression,
    };

    /// Section 7.4. A lambda has no return annotation: its result type comes
    /// from its body, and its parameter types from their annotations or from
    /// the callable type expected where it is written.
    pub const Lambda = struct {
        parameters: []const LambdaParameter,
        body: Body,

        /// A single-expression lambda returns its expression; a block-bodied one
        /// uses `return`. The parser decides by what follows `=>`: a newline
        /// begins a block, and anything else is the one expression.
        pub const Body = union(enum) {
            expression: *const Expression,
            block: Block,
        };
    };

    /// Unlike a named function's parameter, a lambda's annotation is optional
    /// (7.2). `_` discards the argument and binds nothing.
    pub const LambdaParameter = struct {
        name: []const u8,
        name_span: Source.Span,
        annotation: ?TypeExpression,
        /// Section 8.6's `{ (name, age) => ... }`, where the one item a block
        /// receives is a tuple unpacked in the header; null for a plain
        /// parameter.
        pattern: ?Pattern = null,
    };

    /// One `key: value` of a dictionary literal.
    pub const Entry = struct {
        key: *const Expression,
        value: *const Expression,
    };

    /// One piece of an interpolated string: finished text, or an expression
    /// whose displayed value goes there.
    pub const Part = union(enum) {
        text: []const u8,
        expression: *const Expression,
    };

    pub const Index = struct {
        base: *const Expression,
        index: *const Expression,
    };

    pub const Member = struct {
        base: *const Expression,
        name: []const u8,
        name_span: Source.Span,
        /// Section 8.2's `entry.0`, the zero-based position of a tuple member.
        /// Null when the member was written as a name.
        position: ?u32 = null,
    };

    /// Section 6.4's `start..end`, which includes both bounds, or
    /// `start..<end`, which excludes the end. Ranges count upward only.
    pub const Range = struct {
        start: *const Expression,
        end: *const Expression,
        inclusive: bool,
    };

    pub const Unary = struct {
        operator: UnaryOperator,
        operand: *const Expression,
    };

    /// `and` and `or`, which short-circuit and so cannot be ordinary binaries.
    pub const Logical = struct {
        operator: LogicalOperator,
        left: *const Expression,
        right: *const Expression,
    };

    /// A comparison chain such as `0 <= score <= 100`.
    ///
    /// Section 5.2 requires the middle expression to be evaluated once and the
    /// chain to short-circuit as if joined by `and`. Holding the whole chain in
    /// one node is what makes both properties natural rather than something the
    /// evaluator has to reconstruct. `operands.len` is always
    /// `operators.len + 1`.
    pub const Comparison = struct {
        operands: []const *const Expression,
        operators: []const ComparisonOperator,
    };

    pub const Binary = struct {
        operator: BinaryOperator,
        left: *const Expression,
        right: *const Expression,
    };

    pub const Call = struct {
        callee: *const Expression,
        arguments: []const *const Expression,
        /// Section 7.3's named arguments, one per argument and null for a
        /// positional one. Empty when no argument is named.
        names: []const ?ArgumentName = &.{},
        /// Whether the last argument is section 7.4's trailing block, written
        /// after the parentheses. It always fills the final parameter.
        trailing: bool = false,

        pub const ArgumentName = struct {
            text: []const u8,
            span: Source.Span,
        };
    };
};

pub const UnaryOperator = enum {
    negate,
    /// Section 5.2 spells negation `not`. There is no `!` alternative.
    not,

    pub fn lexeme(self: UnaryOperator) []const u8 {
        return switch (self) {
            .negate => "-",
            .not => "not",
        };
    }
};

pub const LogicalOperator = enum {
    conjunction,
    disjunction,

    pub fn lexeme(self: LogicalOperator) []const u8 {
        return switch (self) {
            .conjunction => "and",
            .disjunction => "or",
        };
    }
};

pub const ComparisonOperator = enum {
    equal,
    not_equal,
    less,
    less_equal,
    greater,
    greater_equal,

    pub fn lexeme(self: ComparisonOperator) []const u8 {
        return switch (self) {
            .equal => "==",
            .not_equal => "!=",
            .less => "<",
            .less_equal => "<=",
            .greater => ">",
            .greater_equal => ">=",
        };
    }

    /// `==` and `!=` work on any two values of the same type; the rest need
    /// an order, which only numbers have.
    pub fn isEquality(self: ComparisonOperator) bool {
        return self == .equal or self == .not_equal;
    }

    /// Whether the operator holds for a given ordering. Operands that are not
    /// ordered at all, meaning a NaN is involved, never reach here: section 5.3
    /// gives them IEEE behavior, which the interpreter applies first.
    pub fn holds(self: ComparisonOperator, order: std.math.Order) bool {
        return switch (self) {
            .equal => order == .eq,
            .not_equal => order != .eq,
            .less => order == .lt,
            .less_equal => order != .gt,
            .greater => order == .gt,
            .greater_equal => order != .lt,
        };
    }
};

/// The arithmetic set from section 5.3. Comparison and the word operators arrive
/// with the statement slice, which is the first place a `Bool` is useful.
pub const BinaryOperator = enum {
    add,
    subtract,
    multiply,
    /// `/`, which always produces a `Float`.
    divide,
    /// `//`, which rounds toward negative infinity.
    floor_divide,
    /// `%`, paired with floor division by the law `a == (a // b) * b + (a % b)`.
    remainder,
    /// `**`, which gives an `Int` for two `Int`s and associates right to left.
    power,

    pub fn lexeme(self: BinaryOperator) []const u8 {
        return switch (self) {
            .add => "+",
            .subtract => "-",
            .multiply => "*",
            .divide => "/",
            .floor_divide => "//",
            .remainder => "%",
            .power => "**",
        };
    }

    /// How the operator reads in a diagnostic, in the user's vocabulary.
    pub fn describe(self: BinaryOperator) []const u8 {
        return switch (self) {
            .add => "addition",
            .subtract => "subtraction",
            .multiply => "multiplication",
            .divide => "division",
            .floor_divide => "floor division",
            .remainder => "remainder",
            .power => "exponentiation",
        };
    }
};
